### PR TITLE
Adding cluster creation controls

### PR DESCRIPTION
**Pull request by xavierclements**
Update to the PR that went stale last year - adjusted for API changes.

Added in a wrapper for the POST calls as most shared similarities. TODO: Same for the GET.

### CODE REVIEW

**Comment by xavierclements** [comment]
eesh. forgot to sbt-test. will fix

**Comment by xavierclements** [comment]
@brkyvz Morning. Just trying to figure out why the functionality I added works but the tests are failing. As far as I can tell looks like serialisation of the response is the problem:

```
[info] com.fasterxml.jackson.databind.JsonMappingException: Can not deserialize instance of scala.collection.Seq out of START_OBJECT token
[info]  at [Source: {"id":""}; line: 1, column: 1]
```

Any help?

**Comment by brkyvz** [comment]
@xavierclements I'll take a look at this ASAP. Thanks for providing all this functionality, you've been very helpful!

**Comment by xavierclements** [comment]
@brkyvz cheers. it'll be something embarrassingly simple I bet.

**Comment by xavierclements** [comment]
@brkyvz  just fixed up formatting. tests will still fail.

**Comment by xavierclements** [comment]
@brkyvz had a chance to look at this bug yet? I'm still hitting walls.

File: src/main/scala/sbtdatabricks/DatabricksHttp.scala [hunk 1]
```diff
@@ -82,8 +82,77 @@ class DatabricksHttp(
...
 
+  /*
+   * Check cluster
+   * @param clusterId of the desired cluster
+   * @return cluster 
+   */
+
+  private[sbtdatabricks] def checkCluster(clusterId: String): Option[Cluster] = {
+    val msg = s"Obtaining status information for Cluster Id $clusterId"
+    outputStream.println(msg)
```
Comment (brkyvz) [inline src/main/scala/sbtdatabricks/DatabricksHttp.scala]: would you just like to inline the message and the `println`?

File: src/main/scala/sbtdatabricks/DatabricksHttp.scala [hunk 2]
```diff
@@ -82,8 +82,77 @@ class DatabricksHttp(
...
+    val cluster = fetchClusters.filter(_.id == clusterId)
```
Comment (brkyvz) [inline src/main/scala/sbtdatabricks/DatabricksHttp.scala]: How about leaving this method simply as:
```scala
val cluster = fetchClusters.find(_.id == clusterId)
```

File: src/main/scala/sbtdatabricks/DatabricksHttp.scala [hunk 3]
```diff
@@ -82,8 +82,77 @@ class DatabricksHttp(
     if (entity == null) null else EntityUtils.toString(entity)
   }
 
+  /*
+   * Check cluster
+   * @param clusterId of the desired cluster
+   * @return cluster 
+   */
+
```
Comment (brkyvz) [inline src/main/scala/sbtdatabricks/DatabricksHttp.scala]: nit: remove new line

File: src/main/scala/sbtdatabricks/DatabricksHttp.scala [hunk 4]
```diff
@@ -82,8 +82,77 @@ class DatabricksHttp(
     if (entity == null) null else EntityUtils.toString(entity)
   }
 
+  /*
+   * Check cluster
+   * @param clusterId of the desired cluster
+   * @return cluster
```
Comment (brkyvz) [inline src/main/scala/sbtdatabricks/DatabricksHttp.scala]: `cluster the cluster information if it exists`

File: src/main/scala/sbtdatabricks/DatabricksHttp.scala [hunk 5]
```diff
@@ -82,8 +82,77 @@ class DatabricksHttp(
     if (entity == null) null else EntityUtils.toString(entity)
   }
 
+  /*
+   * Check cluster
+   * @param clusterId of the desired cluster
```
Comment (brkyvz) [inline src/main/scala/sbtdatabricks/DatabricksHttp.scala]: `clusterId id of the desired cluster`

File: src/main/scala/sbtdatabricks/DatabricksHttp.scala [hunk 6]
```diff
@@ -82,8 +82,77 @@ class DatabricksHttp(
...
+   * @return cluster 
+   */
+
+  private[sbtdatabricks] def checkCluster(clusterId: String): Option[Cluster] = {
+    val msg = s"Obtaining status information for Cluster Id $clusterId"
+    outputStream.println(msg)
+    val cluster = fetchClusters.filter(_.id == clusterId)
+    cluster.isEmpty match {
```
Comment (brkyvz) [inline src/main/scala/sbtdatabricks/DatabricksHttp.scala]: delete these parts at is it is not very useful to have it in here.

File: src/main/scala/sbtdatabricks/DatabricksHttp.scala [hunk 7]
```diff
@@ -82,8 +82,77 @@ class DatabricksHttp(
     if (entity == null) null else EntityUtils.toString(entity)
   }
 
+  /*
+   * Check cluster
+   * @param clusterId of the desired cluster
+   * @return cluster 
+   */
+
+  private[sbtdatabricks] def checkCluster(clusterId: String): Option[Cluster] = {
```
Comment (brkyvz) [inline src/main/scala/sbtdatabricks/DatabricksHttp.scala]: rename to `getCluster`?

File: src/main/scala/sbtdatabricks/DatabricksHttp.scala [hunk 8]
```diff
@@ -82,8 +82,77 @@ class DatabricksHttp(
     if (entity == null) null else EntityUtils.toString(entity)
   }
 
+  /*
```
Comment (brkyvz) [inline src/main/scala/sbtdatabricks/DatabricksHttp.scala]: `/**`

File: src/main/scala/sbtdatabricks/DatabricksHttp.scala [hunk 9]
```diff
@@ -82,8 +82,77 @@ class DatabricksHttp(
...
+      useSpot: Boolean,
+      sparkVersion: String,
+      zoneId: String): ClusterId = {
+    val checkExistingClusters = fetchClusters.filter(_.name == name)
+    if (!checkExistingClusters.isEmpty) {
+      outputStream.println(s"Cluster with name '$name' already exists")
+      ClusterId("")
+    } else {
+      postWrapper[ClusterId](CreateClusterInputV1(name, 
+              numWorkerContainers,
```
Comment (brkyvz) [inline src/main/scala/sbtdatabricks/DatabricksHttp.scala]: okay to inline these parameters instead of having each take a line

File: src/main/scala/sbtdatabricks/DatabricksHttp.scala [hunk 10]
```diff
@@ -82,8 +82,77 @@ class DatabricksHttp(
...
+        None 
+    }
+  }
+
   /**
-   * Upload a jar to Databrics Cloud.
+   * Create a cluster
+   * @param name of the cluster to be created
+   * @param memoryGB size in memory (GB) of the new cluster
+   * @param useSpot if true spot instances used, otherwise on-demand instances
```
Comment (brkyvz) [inline src/main/scala/sbtdatabricks/DatabricksHttp.scala]: `whether to use spot instances`

File: src/main/scala/sbtdatabricks/DatabricksHttp.scala [hunk 11]
```diff
@@ -82,8 +82,77 @@ class DatabricksHttp(
...
+        outputStream.println(s"Cluster with Id '$clusterId' does not exist")
+        None 
+    }
+  }
+
   /**
-   * Upload a jar to Databrics Cloud.
+   * Create a cluster
+   * @param name of the cluster to be created
+   * @param memoryGB size in memory (GB) of the new cluster
```
Comment (brkyvz) [inline src/main/scala/sbtdatabricks/DatabricksHttp.scala]: new API supports `numContainers` I believe. We should try to use that.

File: src/main/scala/sbtdatabricks/DatabricksHttp.scala [hunk 12]
```diff
@@ -82,8 +82,77 @@ class DatabricksHttp(
...
+   * @param useSpot if true spot instances used, otherwise on-demand instances
+   * @return The id of the newly created cluster
+   *
+   */
+  private[sbtdatabricks] def createCluster(
+      name: String,
+      numWorkerContainers: Integer,
+      useSpot: Boolean,
+      sparkVersion: String,
+      zoneId: String): ClusterId = {
```
Comment (brkyvz) [inline src/main/scala/sbtdatabricks/DatabricksHttp.scala]: missing doc. In addition, I think it could be an Option[String]. I don't remember it being a required field

File: src/main/scala/sbtdatabricks/DatabricksHttp.scala (hunk 11 above)
Comment (brkyvz) [inline src/main/scala/sbtdatabricks/DatabricksHttp.scala]: in fact, that's what you use. Could you please update the doc?

File: src/main/scala/sbtdatabricks/DatabricksHttp.scala [hunk 13]
```diff
@@ -82,8 +82,77 @@ class DatabricksHttp(
...
+    }
+  }
+  
+  /**
+   * Delete a cluster
+   * @param cluster the cluster to be deleted
+   * @return The id of the deleted cluster
+   */
+  private[sbtdatabricks] def deleteCluster(
+      cluster: Cluster): ClusterId = {
```
Comment (brkyvz) [inline src/main/scala/sbtdatabricks/DatabricksHttp.scala]: nit: inline parameter

File: src/main/scala/sbtdatabricks/DatabricksHttp.scala [hunk 14]
```diff
@@ -82,8 +82,77 @@ class DatabricksHttp(
...
+  }
+
+  /**
+   * Resize a cluster
+   * @param cluster the cluster to be resized
+   * @return The id of the deleted cluster
+   */
+  private[sbtdatabricks] def resizeCluster(
+      cluster: Cluster,
+      numWorkerContainers: Integer): ClusterId = {
```
Comment (brkyvz) [inline src/main/scala/sbtdatabricks/DatabricksHttp.scala]: missing doc

File: src/main/scala/sbtdatabricks/DatabricksHttp.scala [hunk 15]
```diff
@@ -82,8 +82,77 @@ class DatabricksHttp(
...
+   */
+  private[sbtdatabricks] def deleteCluster(
+      cluster: Cluster): ClusterId = {
+    postWrapper[ClusterId](DeleteClusterInputV1(cluster))
+  }
+
+  /**
+   * Resize a cluster
+   * @param cluster the cluster to be resized
+   * @return The id of the deleted cluster
```
Comment (brkyvz) [inline src/main/scala/sbtdatabricks/DatabricksHttp.scala]: `resized cluster`?

File: src/main/scala/sbtdatabricks/DatabricksHttp.scala [hunk 16]
```diff
@@ -403,6 +446,22 @@ class DatabricksHttp(
...
   }
 
+
+  /**
+   * Wrapper for the basic post commands
+   * @param input case class with the relevant post settings
+   * @return Response case class
+   *
+   */
+  private def postWrapper[T<:Responses: Manifest](input: PostInputs): T = {
```
Comment (brkyvz) [inline src/main/scala/sbtdatabricks/DatabricksHttp.scala]: nit: `[T <: Responses:` (space around `<:`)

File: src/main/scala/sbtdatabricks/DatabricksPlugin.scala [hunk 17]
```diff
@@ -137,6 +145,73 @@ object DatabricksPlugin extends AutoPlugin {
     else set
   }
 
+  
+  @annotation.tailrec
+  private def onClusterStatus(client: DatabricksHttp,
```
Comment (brkyvz) [inline src/main/scala/sbtdatabricks/DatabricksPlugin.scala]:
```
onClusterStatus(
    client: DatabricksHttp,
    clusterId: String,
    ...): Option[ClusterId] = {
```

4 spaces please

File: src/main/scala/sbtdatabricks/DatabricksPlugin.scala [hunk 18]
```diff
@@ -137,6 +145,73 @@ object DatabricksPlugin extends AutoPlugin {
...
+                              clusterId: String, 
+                              desiredStatus: DBCClusterStatus,
+                              pollTime: Long): Option[ClusterId] = {
+    val cluster = client.checkCluster(clusterId).get
```
Comment (brkyvz) [inline src/main/scala/sbtdatabricks/DatabricksPlugin.scala]: if cluster was `None` this will throw an exception. How about something like:
```scala
client.checkCluster(clusterId) match {
  case Some(cluster) if cluster.status == DBCClusterError.status =>
     sys.error(...)
  case Some(cluster) if cluster.status == desiredStatus.status =>
     Some(ClusterId(cluster.id))
  case _ =>
     Thread.sleep(pollTime)
     onClusterStatus(client, clusterId, desiredStatus, pollTime)
}
```

File: src/main/scala/sbtdatabricks/DatabricksPlugin.scala [hunk 19]
```diff
@@ -137,6 +145,73 @@ object DatabricksPlugin extends AutoPlugin {
...
+      case DBCClusterError.status =>
+        None
+      case _ =>
+        Thread sleep pollTime
+        onClusterStatus(client, clusterId, desiredStatus, pollTime)
+    }
+  }
+
+  private lazy val createClusterImpl: Def.Initialize[Task[Seq[ClusterId]]] = Def.task {
+      val client = dbcApiClient.value
```
Comment (brkyvz) [inline src/main/scala/sbtdatabricks/DatabricksPlugin.scala]: nit: 2 space indent please

File: src/main/scala/sbtdatabricks/DatabricksPlugin.scala [hunk 20]
```diff
@@ -137,6 +145,73 @@ object DatabricksPlugin extends AutoPlugin {
...
+      val onClusters = dbcClusters.value
+      val workers = dbcNumWorkerContainers.value
+      val spot = dbcSpotInstance.value
+      val version = dbcSparkVersion.value
+      val zoneId = dbcZoneId.value
+      val pollTime = dbcPollTime.value
+
+      onClusters.flatMap(p => {
```
Comment (brkyvz) [inline src/main/scala/sbtdatabricks/DatabricksPlugin.scala]:
```scala
onClusters.flatMap { p =>
   ...
}
```

File: src/main/scala/sbtdatabricks/DatabricksPlugin.scala [hunk 21]
```diff
@@ -137,6 +145,73 @@ object DatabricksPlugin extends AutoPlugin {
...
+
+    client.foreachCluster(onClusters, allClusters) { confirmedCluster =>
+      val clusterId = client.deleteCluster(confirmedCluster)
+      onClusterStatus(client, clusterId.id, DBCClusterTerminated, pollTime)
+      clusterIds :+ clusterId
+    }
+    clusterIds
+  }
+
+  // ADD check for cluster status as well - add tests.
```
Comment (brkyvz) [inline src/main/scala/sbtdatabricks/DatabricksPlugin.scala]: `TODO`?

File: src/main/scala/sbtdatabricks/DatabricksPlugin.scala [hunk 22]
```diff
@@ -216,17 +291,18 @@ object DatabricksPlugin extends AutoPlugin {
...
       val contextStatus = client.checkContext(contextId, cluster)
       contextStatus.status match {
         case DBCContextRunning.status => Some(contextId)
         case DBCContextError.status =>
-          client.destroyContext(contextId, cluster)
+          client.destroyContext(cluster, contextId)
           None
         case _ =>
-          Thread sleep 500
+          Thread sleep pollTime
```
Comment (brkyvz) [inline src/main/scala/sbtdatabricks/DatabricksPlugin.scala]: `.sleep(pollTime)`

File: src/main/scala/sbtdatabricks/DatabricksPlugin.scala [hunk 23]
```diff
@@ -243,10 +319,10 @@ object DatabricksPlugin extends AutoPlugin {
           Some(commandId)
         case DBCCommandError.status =>
           client.cancelCommand(cluster, contextId, commandId)
-          client.destroyContext(contextId, cluster)
+          client.destroyContext(cluster, contextId)
           None
         case _ =>
-          Thread sleep 3000
+          Thread sleep pollTime
```
Comment (brkyvz) [inline src/main/scala/sbtdatabricks/DatabricksPlugin.scala]: ditto

File: src/main/scala/sbtdatabricks/util/PostInputs.scala [hunk 24]
```diff
@@ -0,0 +1,137 @@
...
+
+import sbtdatabricks.util.requests._
+import DBApiEndpoints._
+
+private[sbtdatabricks] trait PostInputs {
+  def initialMessage: String
+  val requestCC: DBApiRequest
+  val dbAPIEndPoint: String
+}
+
```
Comment (brkyvz) [inline src/main/scala/sbtdatabricks/util/PostInputs.scala]: nit: delete line

File: src/main/scala/sbtdatabricks/util/PostInputs.scala [hunk 25]
```diff
@@ -0,0 +1,137 @@
...
+
+private[sbtdatabricks] case class CreateContextInputV1(
+  language: DBCExecutionLanguage,
+  cluster: Cluster
+) extends PostInputs {
```
Comment (brkyvz) [inline src/main/scala/sbtdatabricks/util/PostInputs.scala]: nit: move up one line

File: src/main/scala/sbtdatabricks/util/PostInputs.scala [hunk 26]
```diff
@@ -0,0 +1,137 @@
...
+
+  val requestCC = CreateContextRequestV1(language.is, cluster.id)
+
+  val dbAPIEndPoint = CONTEXT_CREATE
+}
+
+private[sbtdatabricks] case class DestroyContextInputV1(
+  cluster: Cluster,
+  contextId: ContextId
+) extends PostInputs {
```
Comment (brkyvz) [inline src/main/scala/sbtdatabricks/util/PostInputs.scala]: nit: move up one line

File: src/main/scala/sbtdatabricks/util/PostInputs.scala [hunk 27]
```diff
@@ -0,0 +1,137 @@
...
+  val requestCC = DestroyContextRequestV1(cluster.id, contextId.id)
+
+  val dbAPIEndPoint = CONTEXT_DESTROY
+}
+
+private[sbtdatabricks] case class CancelCommandInputV1(
+  cluster: Cluster,
+  contextId: ContextId,
+  commandId: CommandId
+) extends PostInputs {
```
Comment (brkyvz) [inline src/main/scala/sbtdatabricks/util/PostInputs.scala]: nit: move up one line

File: src/main/scala/sbtdatabricks/util/PostInputs.scala [hunk 28]
```diff
@@ -0,0 +1,137 @@
...
+
+  val requestCC = CancelCommandRequestV1(cluster.id, contextId.id, commandId.id)
+
+  val dbAPIEndPoint = COMMAND_CANCEL
+}
+
+private[sbtdatabricks] case class LibraryAttachInputV1(
+  library: UploadedLibrary,
+  cluster: Cluster
+) extends PostInputs {
```
Comment (brkyvz) [inline src/main/scala/sbtdatabricks/util/PostInputs.scala]: nit: move up one line for all following as well please

File: src/main/scala/sbtdatabricks/util/PostInputs.scala [hunk 29]
```diff
@@ -0,0 +1,137 @@
...
+  useSpot: Boolean,
+  sparkVersion: String,
+  zoneId: String
+) extends PostInputs {
+  def initialMessage: String = {
+    s"Creating cluster: ${name}"
+  }
+
+  val requestCC = CreateClusterRequestV1(name, 
+           numWorkerContainers,
```
Comment (brkyvz) [inline src/main/scala/sbtdatabricks/util/PostInputs.scala]: feel free to put multiple on the same line

File: src/main/scala/sbtdatabricks/util/PostInputs.scala [hunk 30]
```diff
@@ -0,0 +1,137 @@
...
+ */
+
+package sbtdatabricks
+
+import sbtdatabricks.util.requests._
+import DBApiEndpoints._
+
+private[sbtdatabricks] trait PostInputs {
+  def initialMessage: String
+  val requestCC: DBApiRequest
```
Comment (brkyvz) [inline src/main/scala/sbtdatabricks/util/PostInputs.scala]: these could also be `def`.

File: src/main/scala/sbtdatabricks/util/PostInputs.scala [hunk 31]
```diff
@@ -0,0 +1,137 @@
...
+
+private[sbtdatabricks] case class CreateContextInputV1(
+  language: DBCExecutionLanguage,
+  cluster: Cluster
+) extends PostInputs {
+  def initialMessage: String = {
+    s"Creating '${language.is}' execution context on cluster '${cluster.name}'"
+  }
+
+  val requestCC = CreateContextRequestV1(language.is, cluster.id)
```
Comment (brkyvz) [inline src/main/scala/sbtdatabricks/util/PostInputs.scala]: all of these need `override`

**Comment by brkyvz** [comment]
@xavierclements Thank you very much for the PR once again. I made a quick pass. I have several comments on style and some cleanup suggestions.
Regarding your exception, it's like you're either wrapping something in a `Seq` when you don't need to, or actually more probably, you're not wrapping a `response` in a `Seq`.

File: src/main/scala/sbtdatabricks/DatabricksHttp.scala (hunk 9 above)
Comment (xavierclements) [inline src/main/scala/sbtdatabricks/DatabricksHttp.scala]: @brkyvz ok. though I'll need to move at least one parameter to a new line due to the 100 character limit

**Comment by xavierclements** [comment]
@brkyvz made the suggested changes. I still need to fix the Test issue though

**Comment by xavierclements** [comment]
@brkyvz turned out that I had: an infinite loop, corner case for cluster creation not handled AND I wasn't providing responses for each client call.

A year later...the tests pass.

**Comment by brkyvz** [comment]
Awesome news @xavierclements !

**Comment by xavierclements** [comment]
@brkyvz any chance of running your eye over the PR again?

**Comment by xavierclements** [comment]
@brkyvz I'm pestering. I know I'm pestering. I'm trying not to pester...but I am pestering. I just want to avoid another year long pause on this PR. Cheers, Xavier the Pest.

**Comment by brkyvz** [comment]
@xavierclements Very very sorry for not getting back to you. I was on vacation last week. Let me take a look. Once again, I apologize :(((

File: src/main/scala/sbtdatabricks/DatabricksHttp.scala [hunk 32]
```diff
@@ -83,7 +83,64 @@ class DatabricksHttp(
...
+  private[sbtdatabricks] def createCluster(
+      name: String,
+      numWorkerContainers: Integer,
+      useSpot: Boolean,
+      sparkVersion: String,
+      zoneId: Option[String]): ClusterId = {
+    val checkExistingClusters = fetchClusters.filter(_.name == name)
+    if (!checkExistingClusters.isEmpty) {
+      outputStream.println(s"Cluster with name '$name' already exists")
+      ClusterId("")
```
Comment (brkyvz) [inline src/main/scala/sbtdatabricks/DatabricksHttp.scala]: why aren't you just returning the correct cluster id?

**Comment by brkyvz** [comment]
@xavierclements Since DB API 2.0 is out, I was thinking of a refactoring available [here](https://github.com/databricks/sbt-databricks/pull/31/files#diff-83f0b5ea7c564d0b3fcebf0813287f31R51). What do you think? I imagine it may be better to support 2.0 and remove all the, "Enter api/1.2" in the URL. It should all be transparent. You wouldn't need to know if you're using 1.2 or 2.0 as long as it works.

In addition, I have concerns regarding the usability of the create cluster api. Would people always want the exact same cluster? What do you think about making `dbcCreateCluster` a command instead of just a task? If that sbt part is too hard, I can add that feature, but I just would love to hear your thoughts.

**Comment by xavierclements** [comment]
@brkyvz you were on holiday? sorry mate. awful move on my part. apologies. 

As for the API refactoring and Task to Command aspect I'll have a go at it and get back to you.

**Comment by xavierclements** [comment]
@brkyvz just rebased the branch, nothing more. 

That refactoring is much neater - improves on the `postWrapper` function in this PR. Given you already have the work would it be better to merge this PR and then create a new PR with the refactoring?

As for the `dbcCreateCluster` task to command change, I'm not sure I get it. I can run `sbt "set dbcZoneId := \"ap-southeast-2c\"" ...(OTHER PARAMETERS)... dbcCreateCluster` so that a new cluster can be created. The same cluster does not have to be created each time.

File: src/main/scala/sbtdatabricks/DatabricksHttp.scala (hunk 32 above)
Comment (xavierclements) [inline src/main/scala/sbtdatabricks/DatabricksHttp.scala]: @brkyvz updated by c8aef62

**Comment by xavierclements** [comment]
@brkyvz hi again. Any thoughts as to my last comment?

**Comment by brkyvz** [comment]
@xavierclements I haven't forgotten about you! Could you please do a rebase (squash) so that I can merge this in as a single commit?

Thanks!

**Comment by xavierclements** [comment]
@brkyvz hey mate. no worries. rebased previously but obviously forgot to do so after the new commits. alas.

**Comment by xavierclements** [comment]
@brkyvz Hey Burak. Are you able to merge this?

**Comment by brkyvz** [comment]
Merged! Thanks

**Comment by xavierclements** [comment]
@brkyvz it's over! It's finally over!!! Cheers for all the help and patience Burak!

**Comment by brkyvz** [comment]
@xavierclements Thank you for your patience! I'm sorry I didn't have time to get to it in a long time.
